Make DataSource independent of query execution (not extend Operator)

Jackie-Jiang opened the pull request:
## Description
- Make `DataSource` independent of the query execution (not extend `Operator`).
- Add interfaces for forward index.
- Make `ForwardIndexReaderContext` close-able so that resources (e.g. off-heap buffer) can be released.
- In `ForwardIndexReader', always read with `ForwardIndexReaderContext` and close it when the reading is done to prevent resource leak.
- Remove `SingleValueSet` and `MultiValueSet` and change caller logic to directly use forward index and prevent resource leak.
- Make `DataFetcher` directly use forward index and handle the type conversion within the class. Similar to Dictionary convention, type conversion among INT, LONG, FLOAT, DOUBLE, STRING, and between STRING and BYTES are supported.

Motivation:
Currently `DataSource` is modeled as an `Operator`, where values are returned as a block complying with the Operator interface.
This is confusing because of the following reasons:
- The block contains all the documents in the segment, instead of a block of at most 10000 documents as in the `Projection` layer.
- The values are always fetched with their document ids, instead of fetched as a block. Currently `BlockValSet` interface has 2 sets of APIs because of this, which is confusing.
- The `BlockValSet` returned by the `DataSource` is not really a value set, but a value reader on top of the forward index.
- Extra `BlockMetadata` has to be maintained which can cause unexpected problems (e.g. the issue fixed in #5619)

TODO:
There are 2 remaining places where `ForwardIndexReaderContext` needs to be closed:
- `ScanBasedDocIdIterator`
- `DataFetcher`

New inheritance hierarchy for the forward index classes:
- `ForwardIndexReader`
  - Dictionary-encoded unsorted SV: `FixedBitSVForwardIndexReader`
  - Dictionary-encoded sorted SV: `SortedIndexReaderImpl`
  - Dictionary-encoded MV: `FixedBitMVForwardIndexReader`
  - Raw fixed-byte SV: `FixedByteChunkSVForwardIndexReader`
  - Raw var-byte SV: `VarByteChunkSVForwardIndexReader`
- `MutableForwardIndex` extends `ForwardIndexReader`
  - Fixed-byte SV (for both dictionary-encoded and raw): `FixedByteSVMutableForwardIndex`
  - Var-byte SV: `VarByteSVMutableForwardIndex`
  - Fixed-byte MV (for both dictionary-encoded and raw): `FixedByteMVMutableForwardIndex`
- `ForwardIndexCreator`
  - Dictionary-encoded unsorted SV: `SingleValueUnsortedForwardIndexCreator`
  - Dictionary-encoded sorted SV: `SingleValueSortedForwardIndexCreator`
  - Dictionary-encoded MV: `MultiValueUnsortedForwardIndexCreator`
  - Raw fixed-byte SV: `SingleValueFixedByteRawIndexCreator`
  - Raw var-byte SV: `SingleValueVarByteRawIndexCreator`

mayankshriv commented on pinot-core/src/main/java/org/apache/pinot/core/common/ColumnValueReader.java:
```diff
@@ -0,0 +1,248 @@
+  public ColumnValueReader(ForwardIndexReader forwardIndexReader) {
+    _forwardIndexReader = forwardIndexReader;
+    _valueType = forwardIndexReader.getValueType();
+    // TODO: Figure out a way to close the reader context. Currently it can cause direct memory leak.
```
This should not be a TODO, but fixed in this PR itself? Or are you saying it happens in some corner case? If so, please describe when that might happen.

Jackie-Jiang replied:
This is the current behavior. The fix is not very straight forward, so I don't want to mix this fix with the refactor PR. Will try to fix it in the next PR.

mayankshriv commented on pinot-core/src/main/java/org/apache/pinot/core/common/ColumnValueReader.java:
```diff
@@ -0,0 +1,248 @@
+ */
+@SuppressWarnings({"rawtypes", "unchecked"})
+public final class ColumnValueReader {
+  private final ForwardIndexReader _forwardIndexReader;
```
Should this extend forwardIndexReader, as it seems to be pass-through for several api's, and providing additional functionality?

Jackie-Jiang replied:
This is a wrapper class on top of the forward index to help manage the reader context. It replaces the current SingleValueReader and MultiValueReader. It should not implement ForwardIndexReader but access forward index.

mayankshriv commented on pinot-core/src/main/java/org/apache/pinot/core/common/ColumnValueReader.java:
```diff
@@ -0,0 +1,248 @@
+   */
+  public void getLongValues(int[] docIds, int length, long[] valueBuffer) {
+    switch (_valueType) {
+      case INT:
```
Does it make sense to read strings as numbers (of course, assuming that these are numbers stored as strings)? Or that use case does not make sense?

Jackie-Jiang replied:
We can try to cast all types, but could be costly (there is a TODO on line 126).
Again, this PR only involves clean up and refactor without any functionality change. We can add that in a separate PR if required.

mayankshriv commented on pinot-core/src/main/java/org/apache/pinot/core/io/readerwriter/ForwardIndexReaderWriter.java:
```diff
@@ -16,11 +16,29 @@
+/**
+ * Interface for forward index reader-writer.
+ */
+public interface ForwardIndexReaderWriter extends ForwardIndexReader<ReaderContext>, ForwardIndexWriter {
```
I am confused by this change:
* Why is this diff showing as move of an unrelated class?
* The name of the interface is `ForwardIndexReaderWriter`, but it has no read or write api's?

Jackie-Jiang replied:
1. Somehow git identifies them as a move. It is supposed to replace `BaseSingleColumnSingleValueReaderWriter` and `BaseSingleColumnMultiValueReaderWriter`
2. It extends `ForwardIndexReader` and `ForwardIndexWriter` which contains all the read and write api's

mayankshriv commented on pinot-core/src/main/java/org/apache/pinot/core/io/readerwriter/impl/FixedByteMVForwardIndexReaderWriter.java:
```diff
@@ -221,102 +222,63 @@ private int updateHeader(int row, int numValues) {
   @Override
-  public void setIntArray(int row, int[] intArray) {
-    int newStartIndex = updateHeader(row, intArray.length);
+  public void setIntArray(int docId, int[] intArray) {
```
Is the caller still expected to call these api's with incremental docId? If so, perhaps get rid of the docId arg, and write at current cursor position?

Jackie-Jiang replied:
Good point. Didn't realize we have such assumption. Will update the PR.

Jackie-Jiang replied:
For ReaderWriter (CONSUMING segment), we still need the DocId because of the aggregate metrics where values are updated in-place.
For Writer, we have the assumption that the value has to be indexed one by one.
Updated the PR to have separate APIs for ReaderWriter and Writer, and also added a TODO to merge ForwardIndexWriter into ForwardIndexCreator.

mayankshriv commented on pinot-core/src/main/java/org/apache/pinot/core/io/writer/ForwardIndexWriter.java:
```diff
@@ -0,0 +1,148 @@
+   * @param docId Document id
+   * @param value Value to write
+   */
+  default void setInt(int docId, int value) {
```
If this is a new interface, why provide `default` here (as opposed to an abstract base class)?

Jackie-Jiang replied:
The difference between default methods in interface vs an common abstract base class:
- A class cannot extend multiple base classes but can implement multiple interfaces
- The interface extending this (e.g. ForwardIndexReaderWriter) can automatically get the default implementation
- IMO this is more readable than a common base class that all classes extend

siddharthteotia commented:
@Jackie-Jiang , It would make the review easy if you can update the PR with the new inheritance hierarchy for the new readers and writers and give some insight into why some of the reader writers have been deleted.

siddharthteotia commented:
> @Jackie-Jiang , It would make the review easy if you can update the PR with the new inheritance hierarchy for the new readers and writers and give some insight into why some of the reader writers have been deleted.

@Jackie-Jiang, it will be useful to add the above info

Jackie-Jiang commented:
@mayankshriv @siddharthteotia Addressed the comments and added the inheritance hierarchy

siddharthteotia commented on pinot-core/src/main/java/org/apache/pinot/core/common/ColumnValueReader.java:
```diff
@@ -0,0 +1,248 @@
+
+
+/**
+ * ColumnValueReader is a wrapper class on top of the forward index reader to read values from it.
```
Why do we need this facade over ForwardIndexReader?

Also, this is misleading since ColumnValue implies the actual value in the column whereas if it is passing the call to ForwardIndex, then for dictionary encoded columns you are essentially reading the dictionaryIds which will then be used by DataFetcher to fetch actual values from the dictionary. 

DataFetcher takes care of reading the forward index and if there is dictionary, it then reads the dictionary. So DataFetcher takes care of reading the actual column values in both cases. I don't think we need this. 

DataFetcher should directly use ForwardIndexReader and Dictionary (which it already does)

Having this class also adds to the unclear semantics -- should ForwardIndexReader be enough or should we wrap it around ColumnValueReader? For example, we are using ColumnValueReader in PinotSegmentColumnReader but directly using ForwardIndexReader in IndexSegmentUtils

siddharthteotia commented on pinot-core/src/main/java/org/apache/pinot/core/common/DataFetcher.java:
```diff
@@ -21,28 +21,24 @@
- * these instances. DataFetcher can be used by both selection, aggregation and group-by data fetching process, reducing
- * duplicate codes and garbage collection.
+ * DataFetcher is a higher level abstraction for data fetching. Given the DataSource, DataFetcher can manage the
+ * ColumnValueReader and Dictionary for the column, preventing redundant construction for these instances. DataFetcher
```
IMO, we should directly use the ForwardIndexReader here

siddharthteotia commented on pinot-core/src/main/java/org/apache/pinot/core/io/reader/impl/BaseChunkSVForwardIndexReader.java:
```diff
@@ -16,27 +16,24 @@
- * Abstract class implementation for {@link BaseSingleColumnSingleValueReader}.
- * Base class for the fixed and variable byte reader implementations.
- *
+ * Base implementation for chunk based single-value forward index reader.
```
Consider changing to - `Base implementation for chunk based single-value raw forward index reader.`

siddharthteotia commented on pinot-core/src/main/java/org/apache/pinot/core/io/reader/impl/FixedByteChunkSVForwardIndexReader.java:
```diff
@@ -16,123 +16,117 @@
 /**
- * Reader class for data written out by {@link FixedByteChunkSingleValueWriter}.
- * For data layout, please refer to the documentation for {@link FixedByteChunkSingleValueReader}
+ * Reader class for data written out by {@link FixedByteChunkSVForwardIndexWriter}.
```
For completeness, let's also say raw forward index reader in the javadoc

siddharthteotia commented on pinot-core/src/main/java/org/apache/pinot/core/io/writer/ForwardIndexWriter.java:
```diff
@@ -0,0 +1,138 @@
+
+/**
+ * Interface for forward index writer.
+ * <p>TODO: Remove this interface and merge the writer implementations into the ForwardIndexCreator implementations.
```
This doesn't sound good. Why are we adding it now if we already know it will be deleted in the future?

On that note, why do we need ForwardIndexCreator and ForwardIndexWriter? We might as well use the writer directly. The segment generation code instantiates creators -> creators instantiate writer. Can we just use writer or one of them whichever is suitable?

siddharthteotia commented on pinot-core/src/main/java/org/apache/pinot/core/io/writer/impl/BaseChunkSVForwardIndexWriter.java:
```diff
@@ -27,21 +27,20 @@
 /**
- * Abstract implementation for {@link SingleColumnSingleValueWriter}
- * Base class for fixed and variable byte writer implementations.
+ * Base implementation for chunk based single-value forward index writer.
```
Consider adding "raw"  -- `Base implementation for chunk based single-value raw forward index writer.`

siddharthteotia commented on pinot-core/src/main/java/org/apache/pinot/core/io/writer/impl/FixedBitSVForwardIndexWriter.java:
```diff
@@ -0,0 +1,59 @@
+ * Bit-compressed dictionary-encoded forward index writer for single-value columns. The values written are dictionary
+ * ids.
+ */
+public class FixedBitSVForwardIndexWriter implements ForwardIndexWriter {
```
Why not just change the name of existing file FixedBitSingleValueWriter and remove all the unsupported operations. ? You are already doing that for FixedBitMVForwardIndexWriter which was originally FixedBitMultiValueWriter

Jackie-Jiang replied:
I don't quite follow. `FixedBitSingleValueWriter` is already renamed to `FixedBitSVForwardIndexWriter`

siddharthteotia replied:
I was suggesting that instead of deleting FixedBitSingleValueReader and creating FixedBitSVForwardIndexReader, you can just rename the file  (and do the cleanup) since you have followed this process for FixedBitMultiValueReader. You have renamed it instead of deleting and creating a new file. Same is applicable for FixdBitSvForwardIndexWriter.

Jackie-Jiang replied:
That is what I did actually, but git might identify it as a new file because the percentage of change is high (the class itself is fairly small)

siddharthteotia commented on pinot-core/src/main/java/org/apache/pinot/core/segment/creator/ForwardIndexCreator.java:
```diff
@@ -19,7 +19,105 @@
+  }
+
+  /**
+   * MULTI-VALUE COLUMN APIs
```
Add a comment stating only dictionary encoded MV columns are supported

siddharthteotia commented on pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/InvertedIndexReader.java:
```diff
@@ -24,13 +24,14 @@
 
   /**
-   * Get the document ids for a given value
+   * Returns the document ids for the given string representation of the value.
```
Can we add APIs for each raw value type so that this interface can also be used for non dictionary based inverted indexes? With the current change, we can probably still do but the raw value has to be converted to string?

Jackie-Jiang replied:
The value here is from the query, so we always pass it in as string, and the inverted index implementation can parse the string based on the value type. This is similar to how dictionary handles the lookup of user input value (`Dictionary.indexOf(String stringValue)`).

siddharthteotia commented on pinot-core/src/test/java/org/apache/pinot/index/readerwriter/FixedByteChunkSVForwardIndexReaderWriterTest.java:
```diff
@@ -276,56 +257,6 @@ public void testDouble(ChunkCompressorFactory.CompressionType compressionType)
     FileUtils.deleteQuietly(outFileEightByte);
   }
 
-  public void testBytes(ChunkCompressorFactory.CompressionType compressionType)
```
This test was for fixed length bytes which for some reason FixedByteChunk writer also supported.

Jackie-Jiang replied:
It can be supported, but is never used. STRING and BYTES are always handled as var-bytes (we don't have the metadata to denote whether a column is fixed-bytes or var-bytes).

Jackie-Jiang commented:
@mayankshriv @siddharthteotia Initially I wanted to make this PR relatively small and just add the new interfaces, but because of the current code structure, it will somehow pollute the interface (e.g. introduces the `ColumnValueReader` class to replace the current `SingleValueReader` and `MultiValueReader`), and cause confusions for the reviewers.
I updated the PR to directly get the clean interface in one step, and addressed all the problems mentioned in the comments. The code change is much bigger, but the interface should be much cleaner, and easier to review.
